README.md
# Pairing on the BLS12-381 Elliptic Curve

This library is a simple and self-contained Haskell implementation of pairing
operations over the BLS12-381 elliptic curve. It contains prime-extension 
finite-field arithmetic arranged in a tower (Fq1, Fq2, Fq6 and Fq12), the
group law for both curve groups in affine coordinates with the sextic twist, 
and culminates in the Ate pairing via a Miller loop. This pairing is used in 
the IETF CFRG draft specification
[bls-signatures](https://datatracker.ietf.org/doc/draft-irtf-cfrg-bls-signature/)

> **Please note:** This code is for experimental purposes and is not suitable
> for production. While a limited testing infrastructure is provided, the code 
> may be incomplete, inefficient, incorrect and/or insecure. Specifically, 
> both the point arithmetic algorithms within the code and the use of Haskell's
> big integers are not constant-time and thus introduce timing side channels.
>
> This code has not undergone a security audit; use at your own risk.

The primary logic resides in `Crypto/Pairing_bls12381.hs`.


## Quick start

After installing Stack, getting started is extremely simple and fast:

    $ git clone https://github.com/nccgroup/pairing-bls12381.git
    $ cd ./pairing-bls12381
    $ stack test
    
        pairing-bls12381> test (suite: pairing-bls12381-test)
    
        Running Tests
          huTstSmokeTest:  OK (1.99s)
          huTstCurve:      OK (0.76s)
          huTstPairingPts: OK (1.53s)
          huTstPairingMul: OK (2.25s)
          huTstPairingGen: OK (9.59s)
    
        All 5 tests passed (9.59s)

        pairing-bls12381> Test suite pairing-bls12381-test passed

As expected, `stack haddock` will generate local documentation and `stack run`
will execute `app/Main.hs` which performs 100 pairings and 30 pointMuls (as a
demo).


## Example usage

Let's demonstrate the following equality (note the constants shifting positions):

`pairing((12+34)*56*g1, 78*g2) == pairing(78*g1, 12*56*g2) * pairing(78*g1, 34*56*g2)`

... where g1 and g2 are group generators. Below is an example `ghci` interpreter 
session.

    $ ghci Crypto/Pairing_bls12381.hs
    
    *Pairing_bls12381> p_12p34m56 = g1Generator >>= pointMul ((12 + 34) * 56)
    *Pairing_bls12381> q_78 = g2Generator >>= pointMul 78
    *Pairing_bls12381> leftSide = pairing <$> p_12p34m56 <*> q_78 >>= id
    *Pairing_bls12381>
    *Pairing_bls12381> p_78 = g1Generator >>= pointMul 78
    *Pairing_bls12381> q_12m56 = g2Generator >>= pointMul (12 * 56)
    *Pairing_bls12381> q_34m56 = g2Generator >>= pointMul (34 * 56)
    *Pairing_bls12381> pair2 = pairing <$> p_78 <*> q_12m56 >>= id
    *Pairing_bls12381> pair3 = pairing <$> p_78 <*> q_34m56 >>= id
    *Pairing_bls12381> rightSide = (*) <$> pair2 <*> pair3
    *Pairing_bls12381>
    *Pairing_bls12381> (==) <$> leftSide <*> rightSide
    Just True


## Straightforward API

The general flow is to first generate two pairs of points. Each pair consists
of one point from group 1 and another point from group 2. Then, each pair is
supplied to the pairing function and the results are compared for equality. It
is possible to perform interesting operations 'within the scalars' of each
point as well as multiply pairing results as demonstrated by the example above.

Points are typically generated by either using `pointMul` on a scalar and the
desired group generator, or by constructing the point directly with 
`g1Point` or `g2Point`. The returned points are be validated to be on-curve.


`g1Point :: Integer -> Integer -> Maybe (Point Fq1)`    
Given `x` and `y`, construct a valid point contained in G1.

`g2Point :: Integer -> Integer -> Integer -> Integer -> Maybe (Point Fq2)`    
Given `xi`, `x`, `yi` and `y`, construct a valid point contained in G2.

`g1Generator :: Maybe (Point Fq1)`    
The standard generator point for G1.

`g2Generator :: Maybe (Point Fq2)`    
The standard generator point for G2.

`pointMul :: (Field a, Eq a) => Integer -> Point a -> Maybe (Point a)`    
Multiply a positive integer scalar and valid point in either G1 or G2.

`pairing :: Point Fq1 -> Point Fq2 -> Maybe Fq12`    
Pairing calculation for a valid point in G1 and another valid point in G2.

`prime :: Integer`    
The field prime constant used in BLS12-381 is exported for reference.

`order :: Integer`    
The curve order constant of BLS12-381 is exported for reference.

`smokeTest :: Bool`    
A quick test of externally inaccessible functionality; returns success.


## References

* Pairings for beginners, Craig Costello, 
  <http://www.craigcostello.com.au/pairings/PairingsForBeginners.pdf>
* On the Implementation of Pairing-Based Cryptography, Ben Lynn PhD 
  Dissertation, <https://crypto.stanford.edu/pbc/thesis.html>
* BLS12-381: New zk-SNARK Elliptic Curve Construction, Sean Bowe of 
  Zcash, <https://electriccoin.co/blog/new-snark-curve/>
* BLS12-381 For The Rest Of Us, Ben Edgington, <https://hackmd.io/@benjaminion/bls12-381>
* Pairing-Friendly Curves, draft-irtf-cfrg-pairing-friendly-curves-04,
  <https://datatracker.ietf.org/doc/draft-irtf-cfrg-pairing-friendly-curves/>
* Constructing Tower Extensions of Finite Fields for Implementation of Pairing-Based 
  Cryptography, Naomi Benger and Michael Scott, <https://eprint.iacr.org/2009/556.pdf>

Copyright (c) 2020 Eric Schorn, NCC Group Plc; Provided under the BSD-3 License.

ChangeLog.md
# Changelog for pairing-bls12381

## Unreleased changes


test/Test.hs
module Main where


import Control.Monad (join)
import Test.Tasty
import qualified Test.Tasty.HUnit as HU
import Pairing_bls12381


main :: IO ()
main = defaultMain $ testGroup "\nRunning Tests" [huTstSmokeTest, huTstCurve,
  huTstPairingPts, huTstPairingMul, huTstPairingGen]


huTstSmokeTest :: TestTree
huTstSmokeTest = HU.testCase "huTstSmokeTest" $ HU.assertBool "bad SmokeTest" smokeTest


huTstCurve :: TestTree
huTstCurve = HU.testCase "huTstCurve" $
  do
    HU.assertBool "bad g1 pointMul"
      ((g1Generator >>= pointMul 0x49abcbaa08d87d1cba8fd9c0ea04df30b94df934827a7383098ac39e1aafc218) ==
        g1Point 0x019906b4953328ec688ffc9e41ea7d79d295c7de6249eb0397680306c5fe3aa3bbb45324bdbc379e8e4116166f2a0d40
                0x165f11693959e7193af31b99b724f95d7b49baa2394b758c2455ef725f32abd56361e1e151f2bbd7a7efc6f3bb5652d4)
    HU.assertBool "bad g2 pointMul"
      ((g2Generator >>= pointMul 0x49abcbaa08d87d1cba8fd9c0ea04df30b94df934827a7383098ac39e1aafc218) ==
        g2Point 0x02433912fa403e0d19d39c7687eb1041f474a82fdd646b1a35afb4d088f11469467468f1ba16c3e5838503919bdbfa24
                0x14906db96db027e17449a1323198cfccde4d15456ce09f3fef4c7baed5495463b7cc750300e0e2918d5680d97a567122
                0x0e82e52250625e4c0864645fba3b36e24c36dbc3d4bae0ca40b0c28b0e3780bf6b8d022c032727e8195e2a2b547be84b
                0x0f240b0ffee3ac62cd5576f012a92cd78c9ded14c11c7637caff4daf885c9a258783a7aef4dc1815737a5b606e03868e)


huTstPairingPts :: TestTree
huTstPairingPts = HU.testCase "huTstPairingPts" $
  do
    let p1 = g1Point 0x14ade5f375a7e7194bf5244ace032817f63ff854f22654e3c0855511156fe98ba5c092ad9d68011f783e0f418619b2f2
                     0x0b54adeba0fdaeeaec7a7146c738c7ebf3f44158475541427ce976fe020e52dedd97feba12c7272b8f3dc49a7d122f8d
    let q1 = g2Point 0x14ebc73c50a68b2333c169565ab75e35c03caaae6573884db2557a6731eca4c643bf7156cb6a2fee60b079577d4732e6
                     0x64974dc7ff588a0b38e37da6b767b91d068924526dd8698f131b6f19b309065aab861b33411675781d9d42b63935c85
                     0x16048641e0249a5ff6cb1b4a8c9194589148c1903b800752d8a6c09867508952d14fecb015cc4f337372b29bc32b08b1
                     0x172f08bfdf8fca06c461c9978e8d8f7de5126a9ee74d086a82d90545d9180e3f9199a7fd2fbd022a9bcfa833ec321e0a
    let leftSide = pairing <$> p1 <*> q1
    let p2 = g1Point 0x19097e1378ae74067593c7e6b71fc8ec41d11e8e1b67b9ca47475e2eb02d5959cfb4bb09a3cb4c7fc14506629b4176d1
                     0xe41648923fc4a39b9e0e0f3d4b0fdcae45caad07622e2d40e67d2eea13bff92f071c0cf4b6687417efdadfae4fec00a
    let q2 = g2Point 0xda18340291cbdd3cfaf596d62fe88f4cf211e8a2439f736786c12fba4e66cccb9279ef1a814d81785bec778c60f9284
                     0xa1d6930cfb119d6b2c52f8260217bf7859437e6aa8b8ca891ce74886f7c667274507b6f107a0c4c7618fe2f3571fcb2
                     0x9745aac53daa7296110ad436829ec3cfd8fbc83012878608479a11caf2ada57c95dd64fd5fd3d4cf3cdf536defd288
                     0x1831856291ad777b5e8f2eff3c2a4efe23a56f44ad97cda5c77f5a9c51bef1e13691011ab1979568363f5b1fecdd32d6
    let rightSide = pairing <$> p2 <*> q2
    HU.assertBool "bad pairing points" (leftSide == rightSide)


huTstPairingMul :: TestTree
huTstPairingMul = HU.testCase "huTstPairingMul" $
  do
    let p_12p34m56 = g1Generator >>= pointMul ((12 + 34) * 56)
    let q_78 = g2Generator >>= pointMul 78
    let leftSide =   Control.Monad.join (pairing <$> p_12p34m56 <*> q_78)
    let p_78 = g1Generator >>= pointMul 78
    let q_12m56 = g2Generator >>= pointMul (12 * 56)
    let q_34m56 = g2Generator >>= pointMul (34 * 56)
    let pair2 = join (pairing <$> p_78 <*> q_12m56)
    let pair3 = join (pairing <$> p_78 <*> q_34m56)
    let rightSide = (*) <$> pair2 <*> pair3
    HU.assertBool "bad pairing mul" (((==) <$> leftSide <*> rightSide) == Just True)


huTstPairingGen :: TestTree
huTstPairingGen = HU.testCase "huTstPairingGen" $
  do
    let scalar_1 = [100..102]
    let scalar_2 = [200..202]
    let p_g1 = [g1Generator >>= pointMul x | x <- scalar_1]
    let p_g2 = [g2Generator >>= pointMul x | x <- scalar_1]
    let q_g1 = [g1Generator >>= pointMul x | x <- scalar_2]
    let q_g2 = [g2Generator >>= pointMul x | x <- scalar_2]
    let pairing1 = [join (pairing <$> p <*> q) | p <- p_g1, q <- q_g2]
    let pairing2 = [join (pairing <$> q <*> p) | p <- p_g2, q <- q_g1]
    HU.assertBool "bad pairing gen" (all (uncurry (==)) (zip pairing1 pairing2))


Crypto/Pairing_bls12381.hs
{-|
Module      : Pairing_bls12381
Description : Pairing over the BLS12-381 elliptic curve
Copyright   : (c) Eric Schorn, NCC Group Plc, 2020
License     : BSD-3
Maintainer  : eric.schorn@nccgroup.com
Stability   : experimental

Implements the BLS12-381 point generation and pairing calculation per
<https://electriccoin.co/blog/new-snark-curve/>.
This code has no dependencies and utilizes no language options.

This code is for strictly experimental purposes: simplicity and clarity are the
primary goals, so the code may be incomplete, inefficient, incorrect and/or insecure.
Specifically, both the algorithms within the code and (the use of) Haskell's
arbitrary-precision integers are clearly not constant-time and thus introduce timing
side channels. This code has not undergone a security audit; use at your own risk.

Note that field element constructors are not exported. Valid points can either be
constructed directly via the `g1Point` or `g2Point` functions, or they can be
constructed from the `pointMul` function given a scalar and typically either the
`g1Generator` or `g2Generator` points. Points are then used in the `pairing` function
which returns `Fq12` elements that can be tested for equality, shown or used in more
elaborate calculations (involving multiplication) with the exported Fq12 `*` operator
(see the example below).

Coordinates are forced into Zp and Points are internally validated to be on-curve. As
such, the `g1Point`, `g1Generator`, `g2Point`, `g2Generator` and `pointMul` functions
all return `Maybe Points` which will need to be unwrapped prior to use in the
`pairing` function.

__Example usage:__

Demonstrate the following equality (note the constants shifting positions):

@pairing((12+34)*56*g1, 78*g2) == pairing(78*g1, 12*56*g2) * pairing(78*g1, 34*56*g2)@

where g1 and g2 are the standard group generators. Below is an example @ghci@
interpreter session.

@\$ ghci Crypto\/Pairing_bls12381.hs
\
*Pairing_bls12381> p_12p34m56 = g1Generator >>= pointMul ((12 + 34) * 56)
*Pairing_bls12381> q_78 = g2Generator >>= pointMul 78
*Pairing_bls12381> leftSide = pairing \<$\> p_12p34m56 \<*\> q_78 >>= id
*Pairing_bls12381>
*Pairing_bls12381> p_78 = g1Generator >>= pointMul 78
*Pairing_bls12381> q_12m56 = g2Generator >>= pointMul (12 * 56)
*Pairing_bls12381> q_34m56 = g2Generator >>= pointMul (34 * 56)
*Pairing_bls12381> pair2 = pairing \<$\> p_78 \<*\> q_12m56 >>= id
*Pairing_bls12381> pair3 = pairing \<$\> p_78 \<*\> q_34m56 >>= id
*Pairing_bls12381> rightSide = (*) \<$\> pair2 \<*\> pair3
*Pairing_bls12381>
*Pairing_bls12381> (==) \<$\> leftSide \<*\> rightSide
Just True@
-}

module Pairing_bls12381 (g1Point, g2Point, g1Generator, g2Generator, pointMul,
                         pairing, prime, order, smokeTest, Num( (*) ) ) where


import Data.Bits (shiftR)
import Data.List (unfoldr)
import Control.Monad (join)


-- Tower extension fields in t, u, v and w. See https://eprint.iacr.org/2009/556.pdf
newtype Fq1 = Fq1 {t0 :: Integer} deriving (Eq, Show)
data Fq2 = Fq2 {u1 :: Fq1, u0 :: Fq1} deriving (Eq, Show)
data Fq6 = Fq6 {v2 :: Fq2, v1 :: Fq2, v0 :: Fq2} deriving (Eq, Show)
data Fq12 = Fq12 {w1 :: Fq6, w0 :: Fq6} deriving (Eq, Show)


-- | The field prime constant used in BLS12-381 is exported for reference.
prime = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab


-- Fields will implement `Num` along with these two functions
class (Num a) => Field a where
  inv :: a -> a
  mul_nonres :: a -> a


-- Fq1 is 'standard' single-element finite field
instance Num Fq1 where

  (+) (Fq1 a0) (Fq1 b0) = Fq1 ((a0 + b0) `mod` prime)  -- Perf opportunity here..

  (-) (Fq1 a0) (Fq1 b0) = Fq1 ((a0 - b0) `mod` prime)  -- ..here; favoring simplicity

  (*) (Fq1 a0) (Fq1 b0) = Fq1 ((a0 * b0) `mod` prime)

  fromInteger a0 = Fq1 (a0 `mod` prime)

  abs = error "not needed/implemented for Fq1"

  signum = error "not needed/implemented for Fq1"


instance Field Fq1 where

  mul_nonres a0 = a0

  -- All fields inverse (incl 0) arrive here
  inv (Fq1 a0) = if a0 == 0 then error "inv of 0" else Fq1 (beea a0 prime 1 0 prime)


-- Binary Extended Euclidean Algorithm (note that there are no divisions)
-- See: Guide to Elliptic Curve Cryptography by Hankerson, Menezes, and Vanstone
beea :: Integer -> Integer -> Integer -> Integer -> Integer -> Integer
beea u v x1 x2 p
  | not (u > 0 && v > 0) = error "beea operands u,v must be greater than zero"
  | u == 1 = mod x1 p
  | v == 1 = mod x2 p
  | even u = if even x1 then beea (shiftR u 1) v (shiftR x1 1) x2 p
                        else beea (shiftR u 1) v (shiftR (x1 + p) 1) x2 p
  | even v = if even x2 then beea u (shiftR v 1) x1 (shiftR x2 1) p
                        else beea u (shiftR v 1) x1 (shiftR (x2 + p) 1) p
  | u >= v = beea (u - v) v (x1 - x2) x2 p
  | u < v  = beea u (v - u) x1 (x2 - x1) p


-- Fq2 is constructed with Fq1(u) / (u^2 - β) where β = -1.
instance Num Fq2 where

  (+) (Fq2 a1 a0) (Fq2 b1 b0) = Fq2 (a1 + b1) (a0 + b0)

  (-) (Fq2 a1 a0) (Fq2 b1 b0) = Fq2 (a1 - b1) (a0 - b0)

  (*) (Fq2 a1 a0) (Fq2 b1 b0) = Fq2 (a1 * b0 + a0 * b1) (a0 * b0 - a1 * b1)

  fromInteger a0 = Fq2 0 (fromInteger a0)

  abs = error "not needed/implemented for Fq2"

  signum = error "not needed/implemented for Fq2"


instance Field Fq2 where

  mul_nonres (Fq2 a1 a0) = Fq2 (a1 + a0) (a0 - a1)

  inv (Fq2 a1 a0) = Fq2 (-a1 * factor) (a0 * factor)
    where
      factor = inv (a1 * a1 + a0 * a0)


-- Fq6 is constructed with Fq2(v) / (v^3 - ξ) where ξ = u + 1
instance Num Fq6 where

  (+) (Fq6 a2 a1 a0) (Fq6 b2 b1 b0) = Fq6 (a2 + b2) (a1 + b1) (a0 + b0)

  (-) (Fq6 a2 a1 a0) (Fq6 b2 b1 b0) = Fq6 (a2 - b2) (a1 - b1) (a0 - b0)

  (*) (Fq6 a2 a1 a0) (Fq6 b2 b1 b0) = Fq6 t2 (t1 + t4) (t0 + t3)
    where
      t0 = a0 * b0
      t1 = a0 * b1 + a1 * b0
      t2 = a0 * b2 + a1 * b1 + a2 * b0
      t3 = mul_nonres (a1 * b2 + a2 * b1)
      t4 = mul_nonres (a2 * b2)

  fromInteger a0 = Fq6 0 0 (fromInteger a0)

  abs = error "not needed/implemented for Fq6"

  signum = error "not needed/implemented for Fq6"


instance Field Fq6 where

  mul_nonres (Fq6 a2 a1 a0) = Fq6 a1 a0 (mul_nonres a2)

  inv (Fq6 a2 a1 a0) = Fq6 (t2 * factor) (t1 * factor) (t0 * factor)
    where
      t0 = a0 * a0 - mul_nonres (a1 * a2)
      t1 = mul_nonres (a2 * a2) - a0 * a1
      t2 = a1 * a1 - a0 * a2
      factor = inv (a0 * t0 + mul_nonres (a2 * t1) + mul_nonres (a1 * t2))


-- Fq12 is constructed with Fq6(w) / (w^2 - γ) where γ = v
instance Num Fq12 where

  (+) (Fq12 a1 a0) (Fq12 b1 b0) = Fq12 (a1 + b1) (a0 + b0)

  (-) (Fq12 a1 a0) (Fq12 b1 b0) = Fq12 (a1 - b1) (a0 - b0)

  (*) (Fq12 a1 a0) (Fq12 b1 b0) =
    Fq12 (a1 * b0 + a0 * b1) (a0 * b0 + mul_nonres (a1 * b1))

  fromInteger a0 = Fq12 0 (fromInteger a0)

  abs = error "not needed/implemented for Fq12"

  signum = error "not needed/implemented for Fq12"


instance Field Fq12 where

  mul_nonres _ = error "not needed for Fq12"

  inv (Fq12 a1 a0) = Fq12 (-a1 * factor) (a0 * factor)
    where
      factor = inv (a0 * a0 - mul_nonres (a1 * a1))


-- Affine coordinates are used throughout for simplicity
data Point a = Affine {ax :: a, ay :: a}
             | PointAtInfinity deriving (Eq, Show)


-- | The curve order constant of BLS12-381 is exported for reference.
order = 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001


-- | The standard generator point for G1.
g1Generator = Just (Affine (Fq1 0x17f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb)
                           (Fq1 0x08b3f481e3aaa0f1a09e30ed741d8ae4fcf5e095d5d00af600db18cb2c04b3edd03cc744a2888ae40caa232946c5e7e1))


-- | The standard generator point for G2.
g2Generator = Just (Affine (Fq2 (Fq1 0x13e02b6052719f607dacd3a088274f65596bd0d09920b61ab5da61bbdc7f5049334cf11213945d57e5ac7d055d042b7e)
                                (Fq1 0x024aa2b2f08f0a91260805272dc51051c6e47ad4fa403b02b4510b647ae3d1770bac0326a805bbefd48056c8c121bdb8))
                           (Fq2 (Fq1 0x0606c4a02ea734cc32acd2b02bc28b99cb3e287e85a763af267492ab572e99ab3f370d275cec1da1aaa9075ff05f79be)
                                (Fq1 0x0ce5d527727d6e118cc9cdc6da2e351aadfd9baa8cbdd3a76d429a695160d12c923ac9cc3baca289e193548608b82801)))


-- BLS12-381 curve(s) are E1: y^2 = x^3 + 4 and E2: y^2 = x^3 + 4(u+1)
isOnCurve :: (Field a, Eq a) => Point a -> Bool
isOnCurve PointAtInfinity = False
isOnCurve a0 = ay a0^2 == (ax a0^3 + mul_nonres 4)


-- Check subgroup=order membership
isInSubGroup :: (Field a, Eq a) => Point a -> Bool
isInSubGroup p = pointMul order p == Just PointAtInfinity


-- | Given @x@ and @y@, construct a valid point contained in G1.
g1Point :: Integer -> Integer -> Maybe (Point Fq1)
g1Point x y = if isOnCurve candidate && isInSubGroup candidate
              then Just candidate else Nothing
  where
    candidate = Affine (fromInteger x) (fromInteger y)


-- | Given @xi@, @x@, @yi@ and @y@, construct a valid point contained in G2.
g2Point :: Integer -> Integer -> Integer -> Integer -> Maybe (Point Fq2)
g2Point x1 x0 y1 y0 = if isOnCurve candidate && isInSubGroup candidate
                      then Just candidate else Nothing
  where
    candidate = Affine (Fq2 (fromInteger x1) (fromInteger x0))
                       (Fq2 (fromInteger y1) (fromInteger y0))


-- Add affine curve points; handle all corner cases
pointAdd :: (Field a, Eq a) => Point a -> Point a -> Point a
pointAdd PointAtInfinity q = q
pointAdd p PointAtInfinity = p
pointAdd Affine {ax=x1, ay=y1} Affine {ax=x2, ay=y2}
  | x1 == x2 && y1 == y2 = pointDouble Affine {ax=x1, ay=y1}
  | x1 == x2 && y1 /= y2 = PointAtInfinity
  | otherwise = Affine {ax=x3, ay=y3}
  where
    slope = (y2 - y1) * inv (x2 - x1)
    x3 = slope^2 - x1 - x2
    y3 = slope * (x1 - x3) - y1


-- Double affine curve points
pointDouble :: (Field a) => Point a -> Point a
pointDouble PointAtInfinity = PointAtInfinity
pointDouble Affine {ax=x1, ay=y1} = Affine {ax=x3, ay=y3}
  where
    slope = (3 * x1^2) * inv (2 * y1)
    x3 = slope^2 - x1 - x1
    y3 = slope * (x1 - x3) - y1


-- | Multiply a positive integer scalar and valid point in either G1 or G2.
pointMul :: (Field a, Eq a) => Integer -> Point a -> Maybe (Point a)
pointMul scalar base
  | isOnCurve base && scalar > 0 = Just (pointMul' scalar base PointAtInfinity)
  | otherwise = Nothing


-- Double and add helper loop
pointMul' :: (Field a, Eq a) => Integer -> Point a -> Point a -> Point a
pointMul' scalar base accum
  | scalar == 0 = accum
  | odd scalar  = pointMul' (shiftR scalar 1) doubleBase (pointAdd accum base)
  | even scalar = pointMul' (shiftR scalar 1) doubleBase accum
  where
    doubleBase = pointAdd base base


-- Untwist point on E2 for pairing calculation
untwist :: Point Fq2 -> Point Fq12
untwist Affine {ax=x1, ay=y1} = Affine {ax=wideX, ay=wideY}
  where
    root = Fq6 0 1 0
    wsq = inv (Fq12 0 root)
    wcu = inv (Fq12 root 0)
    wideX = Fq12 0 (Fq6 0 0 x1) * wsq
    wideY = Fq12 0 (Fq6 0 0 y1) * wcu


doubleEval :: Point Fq2 -> Point Fq1 -> Fq12
doubleEval r p = fromInteger (t0 (ay p)) - (fromInteger (t0 (ax p)) * slope) - v
  where
    wideR = untwist r
    rx2 = ax wideR^2
    slope = (3 * rx2) * inv (2 * ay wideR)
    v = ay wideR - slope * ax wideR


-- Used in miller loop when current bit index is true
addEval :: Point Fq2 -> Point Fq2 -> Point Fq1 -> Fq12
addEval r q p = if (ax wideR == ax wideQ) && (ay wideR == - ay wideQ)
                then fromInteger (t0 (ax p)) - ax wideR
                else addEval' wideR wideQ p
  where
    wideR = untwist r
    wideQ = untwist q


addEval' :: Point Fq12 -> Point Fq12 -> Point Fq1 -> Fq12
addEval' wideR wideQ p = fromInteger (t0 (ay p)) - (fromInteger (t0 (ax p)) * slope) - v
  where
    slope = (ay wideQ - ay wideR) * inv (ax wideQ - ax wideR)
    v = ((ay wideQ * ax wideR) - (ay wideR * ax wideQ)) * inv (ax wideR - ax wideQ)


-- Classic Miller loop for Ate pairing
miller :: Point Fq1 -> Point Fq2 -> Fq12
miller p q = miller' p q q iterations 1
  where
    iterations = tail $ reverse $  -- list of true/false per bits of operand
      unfoldr (\b -> if b == (0 :: Integer) then Nothing
                     else Just(odd b, shiftR b 1)) 0xd201000000010000


-- Double and add helper for miller
miller' :: Point Fq1 -> Point Fq2 -> Point Fq2 -> [Bool] -> Fq12 -> Fq12
miller' p q r [] result = result
miller' p q r (i:iters) result =
  if i then miller' p q (pointAdd doubleR q) iters (accum * addEval doubleR q p)
       else miller' p q doubleR iters accum
  where
    accum = result^2 * doubleEval r p
    doubleR = pointDouble r


-- Used for the final exponentiation; opportunity for further perf optimization
pow' :: (Field a) => a -> Integer -> a -> a
pow' a0 exp result
  | exp <= 1 = a0
  | even exp = accum^2
  | otherwise = accum^2 * a0
  where accum = pow' a0 (shiftR exp 1) result


-- | Pairing calculation for a valid point in G1 and another valid point in G2.
pairing :: Point Fq1 -> Point Fq2 -> Maybe Fq12
pairing p_g1 q_g2
  | p_g1 == PointAtInfinity || q_g2 == PointAtInfinity = Nothing
  | isOnCurve p_g1 && isInSubGroup p_g1 && isOnCurve q_g2 && isInSubGroup q_g2
      = Just (pow' (miller p_g1 q_g2) (div (prime^12 - 1) order) 1)
  | otherwise = Nothing


-- | A quick test of externally inaccessible functionality; returns success.
smokeTest :: Bool
smokeTest = and $ res1 ++ res2 ++ res6 ++ res12 ++ resMul ++ resOne
  where
    operands  = [100..1100]
    res1  = [(fromInteger x :: Fq1)  * inv (fromInteger x) == 1 | x <- operands]
    res2  = [(fromInteger x :: Fq2)  * inv (fromInteger x) == 1 | x <- operands]
    res6  = [(fromInteger x :: Fq6)  * inv (fromInteger x) == 1 | x <- operands]
    res12 = [(fromInteger x :: Fq12) * inv (fromInteger x) == 1 | x <- operands]
    resMul = [(g1Generator >>= pointMul order) == Just PointAtInfinity]
    p_12p34m56 = g1Generator >>= pointMul ((12 + 34) * 56)
    q_78 = g2Generator >>= pointMul 78
    (Just pair) =  Control.Monad.join (pairing <$> p_12p34m56 <*> q_78)
    resOne = [pow' pair order 0 == 1]  -- confirms pair result is rth root of unity
